marlowe-dashboard-client/src/MainFrame/State.purs

```diff
@@ -45,81 +48,138 @@ mkMainFrame =
 
 initialState :: State
 initialState =
-  { overlay: Nothing
-  , screen: Home
-  , card: Nothing
-  , contactState: Contact.initialState
+  { wallets: empty
+  , newWalletNicknameKey: emptyWalletNicknameKey
+  , subState: Left initialOutsideState
   , contractState: Contract.initialState zero Close
-  , notifications: []
-  , templates: []
-  , contracts: []
+  }
+
+initialOutsideState :: OutsideState
+initialOutsideState =
+  { screen: GenerateWalletScreen
+  , card: Nothing
+  }
+
+mkInsideState :: PubKeyHash -> InsideState
+mkInsideState pubKeyHash =
+  { wallet: pubKeyHash
+  , menuOpen: false
+  , screen: ContractsScreen Running
+  , card: Nothing
   , on: true
   }
 
+emptyWalletNicknameKey :: WalletNicknameKey
+emptyWalletNicknameKey = Tuple "" ""
+
+defaultWalletDetails :: WalletDetails
+defaultWalletDetails = WalletDetails { userHasPickedUp: false }
+
 handleQuery :: forall a m. Query a -> HalogenM State Action ChildSlots Msg m (Maybe a)
 handleQuery (ReceiveWebSocketMessage msg next) = do
   case msg of
-    (WS.ReceiveMessage (Right (ClientMsg on))) -> assign _on on
+    (WS.ReceiveMessage (Right (ClientMsg on))) -> assign (_insideState <<< _on) on
     -- TODO: other matches such as update current slot or apply transaction
     _ -> pure unit
   pure $ Just next
 
 handleAction :: forall m. MonadAff m => Action -> HalogenM State Action ChildSlots Msg m Unit
 handleAction Init = do
-  mCachedContactsJson <- liftEffect $ getItem contactsLocalStorageKey
-  for_ mCachedContactsJson
+  mCachedWalletsJson <- liftEffect $ getItem walletsLocalStorageKey
+  for_ mCachedWalletsJson
     $ \json -> do
         let
-          contacts =
-            runExcept
-              $ do
-                  foreignJson <- parseJSON json
-                  decode foreignJson
-        case contacts of
-          Right cachedContacts -> assign (_contactState <<< _contacts) cachedContacts
+          wallets =
+            runExcept do
```

Comment:
very minor, but the do is not needed anymore

Reply:
FYI `Either` is `Foldable` which means you can use `for_ (runExceptT $ decodeJSON json) \wallets -> assign _wallets cachedWallets` if you want? Personally I think that looks quite nice but it's pretty subjective.